content/curried-varargs.md

```diff
@@ -0,0 +1,271 @@
+  }
+}
+```
+
+However, the Scala compiler is unable to optimize repeated parameters, as a result, the output JavaScript from Scala.js would look like the below code.
+
+``` javascript
+var $$this$1 = $m_Lorg_scalajs_dom_package$().document__Lorg_scalajs_dom_raw_HTMLDocument().createElement("div");
+var this$3 = $m_LScalaFiddle$xml$attributes$();
+var jsx$1 = new $c_sjsr_AnonFunction1().init___sjs_js_Function1((function($this, value) {
+  return (function(x$1$2) {
+    x$1$2.setAttribute("title", value)
+  })
+})(this$3, "my-title"));
+var $$this = $m_Lorg_scalajs_dom_package$().document__Lorg_scalajs_dom_raw_HTMLDocument().createElement("br");
+var array = [jsx$1, "line1", $$this, "line2"];
+var i = 0;
+var len = $uI(array.length);
+while ((i < len)) {
+  var index = i;
+  var arg1 = array[index];
+  if ($is_T(arg1)) {
+    var x2 = $as_T(arg1);
+    $$this$1.appendChild($m_Lorg_scalajs_dom_package$().document__Lorg_scalajs_dom_raw_HTMLDocument().createTextNode(x2))
+  } else if ($uZ((arg1 instanceof $g.Node))) {
+    $$this$1.appendChild(arg1)
+  } else if ($is_F1(arg1)) {
+    var x4 = $as_F1(arg1);
+    x4.apply__O__O($$this$1)
+  } else {
+    throw new $c_s_MatchError().init___O(arg1)
+  };
+  i = ((1 + i) | 0)
+};
+```
+
+Despite of the type safety issue due to the usage of `Any`, the above code are inefficient: 
+
+1. Unnecessary temporary object for the `xml.attributes.title(xml.text("my-title"))`.
+2. Unnecessary temporary `Seq` to hold repeated parameters.
+3. Unnecessary runtime type check for each argument.
+
+The similar issues can be found in many other usage of repeated parameters. For example, Scala string interpolation is inefficient due to its internal vararg function call, unless implementing it in a macro; Scala collection initializers (e.g. `List(1, 2, 3)`) create unnecessary temporary `Seq` before creating the desired collection.
+
+## Design
+
+This proposal introduces a new type `Curried` defined as following:
+
+``` scala
+trait Curried extends Any
+```
+
+When a function call `f(p1, p2, p3, ... pn)` is being type checked, the compiler will firstly look for `apply` method on `f`. If an applicable `apply` method is not found and `f` is a subtype of `Curried`, the compiler will convert the function call to curried form `f.applyBegin.applyNext(p1).applyNext(p2).applyNext(p3) ... .applyNext(pn).applyEnd`, and continue type checking the translated call.
+
+### Expanding sequence argument
+
+Optionally, some arguments to a `Curried` call may be a sequence argument marked as  `_*`. Those are arguments should be translated to `applyNextSeq` calls instead of `applyNext`. For example, `f(p1, s1: _*, p2)` will be translated to the following code.
+
+``` scala
+f.applyBegin
+  .applyNext(p1)
+  .applyNextSeq(s1)
+  .applyNext(p2)
+.applyEnd
+```
+
+Unlike traditional repeated parameters, which restrict the sequence argument at the last position, sequence arguments in a curried call are allowed at any position.
```

Comment:
I don't think this should be included in this proposal. The desugaring here doesn't make it any more or less difficult to support `_*` arguments in the middle than the regular desugaring. Therefore, it should a separate, orthogonal proposal.